File: src/logic/__tests__/build.test.ts
import { describe, expect, it } from 'vitest';
import { gameReducer, createInitialState } from '../gameReducer';
import { GameActionType } from '../../types/game';
import { canBuildOnCurrentSpace } from '../build';

function makeState() {
  return gameReducer(createInitialState(), {
    type: GameActionType.StartGame,
    playerCount: 2,
    names: ['Alice', 'Bob'],
  });
}

describe('canBuildOnCurrentSpace', () => {
  it('returns false before any roll', () => {
    const s = makeState();
    expect(canBuildOnCurrentSpace(s)).toBe(false);
  });

  it('returns false when the player is not on their own buildable property', () => {
    const s = { ...makeState(), dice: [1, 2] as [number, number] };
    expect(canBuildOnCurrentSpace(s)).toBe(false);
  });

  it('returns true on own property below MAX_HOUSES, not mortgaged, not just bought', () => {
    const base = makeState();
    const property = base.board.find((sp) => sp.type === 'property');
    if (!property) throw new Error('no property space');
    const s = {
      ...base,
      dice: [1, 2] as [number, number],
      currentPlayer: 0,
      board: base.board.map((sp) =>
        sp.id === property.id
          ? { ...sp, owner: 0, houses: 0, mortgaged: false }
          : sp,
      ),
      players: base.players.map((p, i) => (i === 0 ? { ...p, position: property.id } : p)),
    };
    expect(canBuildOnCurrentSpace(s)).toBe(true);
  });

  it('returns false on a mortgaged property', () => {
    const base = makeState();
    const property = base.board.find((sp) => sp.type === 'property');
    if (!property) throw new Error('no property space');
    const s = {
      ...base,
      dice: [1, 2] as [number, number],
      currentPlayer: 0,
      board: base.board.map((sp) =>
        sp.id === property.id
          ? { ...sp, owner: 0, houses: 0, mortgaged: true }
          : sp,
      ),
      players: base.players.map((p, i) => (i === 0 ? { ...p, position: property.id } : p)),
    };
    expect(canBuildOnCurrentSpace(s)).toBe(false);
  });

  it('returns false when dice is null even on own property', () => {
    const base = makeState();
    const property = base.board.find((sp) => sp.type === 'property');
    if (!property) throw new Error('no property space');
    const s = {
      ...base,
      dice: null,
      currentPlayer: 0,
      board: base.board.map((sp) =>
        sp.id === property.id
          ? { ...sp, owner: 0, houses: 0, mortgaged: false }
          : sp,
      ),
      players: base.players.map((p, i) => (i === 0 ? { ...p, position: property.id } : p)),
    };
    expect(canBuildOnCurrentSpace(s)).toBe(false);
  });
});

File: backend/internal/usecase/gameusecase/build_test.go
package gameusecase

import (
	"testing"

	"monopoly-game-backend/internal/data"
	"monopoly-game-backend/internal/entity"
)

func buildState(opts ...func(*entity.GameState)) entity.GameState {
	owner := 1
	state := entity.GameState{
		Players: []entity.Player{
			{ID: 0, Position: 0, Properties: []int{}, Money: data.StartingMoney},
			{ID: 1, Position: 1, Properties: []int{1}, Money: data.StartingMoney},
		},
		Board: []entity.Space{
			{ID: 0, Type: entity.SpaceTypeGo},
			{ID: 1, Type: entity.SpaceTypeProperty, Owner: &owner, Houses: 0, HouseCost: []int{50}, Color: strPtr("#8B4513")},
			{ID: 2, Type: entity.SpaceTypeChance},
			{ID: 3, Type: entity.SpaceTypeTax},
		},
		CurrentPlayer: 1,
		Dice:          &[2]int{3, 4},
	}
	for _, fn := range opts {
		fn(&state)
	}
	return state
}

func TestCanBuildOnCurrentSpace_NormalCase(t *testing.T) {
	state := buildState()
	if !CanBuildOnCurrentSpace(state) {
		t.Error("expected true for normal buildable case")
	}
}

func TestCanBuildOnCurrentSpace_RequiresCurrentProperty(t *testing.T) {
	state := buildState(func(s *entity.GameState) {
		s.Players[1].Position = 0
	})
	if CanBuildOnCurrentSpace(state) {
		t.Error("expected false when an eligible owned property is not the current space")
	}
}

func TestCanBuildOnCurrentSpace_NotPropertyType(t *testing.T) {
	state := buildState(func(s *entity.GameState) {
		s.Board[1].Type = entity.SpaceTypeChance
	})
	if CanBuildOnCurrentSpace(state) {
		t.Error("expected false for non-property type")
	}
}

func TestCanBuildOnCurrentSpace_NoDiceRolled(t *testing.T) {
	state := buildState(func(s *entity.GameState) {
		s.Dice = nil
	})
	if CanBuildOnCurrentSpace(state) {
		t.Error("expected false when dice not rolled")
	}
}

func TestCanBuildOnCurrentSpace_NotOwner(t *testing.T) {
	other := 0
	state := buildState(func(s *entity.GameState) {
		s.Board[1].Owner = &other
	})
	if CanBuildOnCurrentSpace(state) {
		t.Error("expected false when not owner")
	}
}

func TestCanBuildOnCurrentSpace_AlreadyMaxHouses(t *testing.T) {
	state := buildState(func(s *entity.GameState) {
		s.Board[1].Houses = data.MaxHouses
	})
	if CanBuildOnCurrentSpace(state) {
		t.Error("expected false when at max houses")
	}
}

func TestCanBuildOnCurrentSpace_Mortgaged(t *testing.T) {
	state := buildState(func(s *entity.GameState) {
		s.Board[1].Mortgaged = true
	})
	if CanBuildOnCurrentSpace(state) {
		t.Error("expected false when mortgaged")
	}
}

func TestCanBuildOnCurrentSpace_JustBoughtThisSpace(t *testing.T) {
	spaceID := 1
	state := buildState(func(s *entity.GameState) {
		s.JustBoughtSpaceID = &spaceID
	})
	if CanBuildOnCurrentSpace(state) {
		t.Error("expected false when just bought this space")
	}
}

func TestCanBuildOnCurrentSpace_JustBoughtDifferentSpace(t *testing.T) {
	spaceID := 99
	state := buildState(func(s *entity.GameState) {
		s.JustBoughtSpaceID = &spaceID
	})
	if !CanBuildOnCurrentSpace(state) {
		t.Error("expected true when just bought a different space")
	}
}

func TestCanBuildOnCurrentSpace_UnaffordableOffCurrentProperty(t *testing.T) {
	state := buildState(func(s *entity.GameState) {
		s.Players[1].Position = 0
		s.Players[1].Money = 0
	})
	if CanBuildOnCurrentSpace(state) {
		t.Error("expected false when an off-current property is unaffordable")
	}
}

func TestCanBuildOnCurrentSpace_NoPlayer(t *testing.T) {
	state := buildState(func(s *entity.GameState) {
		s.Players = []entity.Player{}
		s.CurrentPlayer = 0
	})
	if CanBuildOnCurrentSpace(state) {
		t.Error("expected false when no player")
	}
}

func TestCanBuildOnCurrentSpace_NoEligibleProperty(t *testing.T) {
	state := buildState(func(s *entity.GameState) {
		s.Players[1].Position = 99
		s.Board[1].Owner = nil
	})
	if CanBuildOnCurrentSpace(state) {
		t.Error("expected false when no eligible property exists")
	}
}
